pkg/controller/clusterpool/collections.go

```diff
@@ -584,6 +599,252 @@ func (cds *cdCollection) Delete(c client.Client, cdName string) error {
+			cdc.Status.LastApplyStatus = hivev1.LastApplyBrokenCloud
+			cdc.Status.LastApplyTime = metav1.Now()
+			if err := c.Status().Update(context.TODO(), cdc); err != nil {
+				return err
+			}
+		}
+	}
+	// Notice a CD has finished installing => update the CDC's LastApplyStatus to Success;
+	for _, cd := range cds.Unassigned(false) {
+		if cd.Spec.ClusterPoolRef.CustomizationRef == nil {
+			continue
+		}
+		if cdc, ok := cdcs.byCDCName[cd.Spec.ClusterPoolRef.CustomizationRef.Name]; ok {
+			if cdc.Status.LastApplyStatus != hivev1.LastApplySucceeded {
+				cdc.Status.LastApplyStatus = hivev1.LastApplySucceeded
+				cdc.Status.LastApplyTime = metav1.Now()
+				if err := c.Status().Update(context.TODO(), cdc); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// Update Cluster Pool Inventory Condition
+	message := ""
+	status := corev1.ConditionTrue
+	reason := hivev1.InventoryReasonValid
+	if (len(cdcs.syntax) + len(cdcs.cloud) + len(cdcs.missing)) > 0 {
+		messageByte, err := json.Marshal(cdcs)
+		if err != nil {
+			return err
+		}
+		message = string(messageByte)
+		status = corev1.ConditionFalse
+		reason = hivev1.InventoryReasonInvalid
+	}
+
+	conditions, changed := controllerutils.SetClusterPoolConditionWithChangeCheck(
+		pool.Status.Conditions,
+		hivev1.ClusterPoolInventoryValidCondition,
+		status,
+		reason,
+		message,
+		controllerutils.UpdateConditionIfReasonOrMessageChange,
+	)
+
+	if changed {
+		pool.Status.Conditions = conditions
+		if err := c.Status().Update(context.TODO(), pool); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cdcs *cdcCollection) MarshalJSON() ([]byte, error) {
+	cloud := []string{}
+	for _, cdc := range cdcs.cloud {
+		cloud = append(cloud, cdc.Name)
+	}
+	syntax := []string{}
+	for _, cdc := range cdcs.syntax {
+		syntax = append(syntax, cdc.Name)
+	}
+
+	return json.Marshal(&struct {
+		BrokenByCloud  []string
+		BrokenBySyntax []string
+		Missing        []string
+	}{
+		BrokenByCloud:  cloud,
+		BrokenBySyntax: syntax,
+		Missing:        cdcs.missing,
+	})
+}
+
+func setCustomizationAvailabilityCondition(c client.Client, cdc *hivev1.ClusterDeploymentCustomization, cd *hivev1.ClusterDeployment, logger log.FieldLogger) error {
+	status := corev1.ConditionTrue
+	reason := "Available"
+	message := "available"
+	cdc.Status.ClusterDeploymentRef = nil
+	cdc.Status.ClusterPoolRef = nil
+
+	if cd != nil {
+		status = corev1.ConditionFalse
+		reason = "Reserved"
+		message = "reserved"
+		cdc.Status.ClusterDeploymentRef = &corev1.LocalObjectReference{Name: cd.Name}
+		cdc.Status.ClusterPoolRef = &corev1.LocalObjectReference{Name: cd.Spec.ClusterPoolRef.PoolName}
+	}
+
+	existingCondition := conditionsv1.FindStatusCondition(cdc.Status.Conditions, conditionsv1.ConditionAvailable)
+	if existingCondition == nil || existingCondition.Reason != reason || existingCondition.Message != message {
+		conditionsv1.SetStatusConditionNoHeartbeat(&cdc.Status.Conditions, conditionsv1.Condition{
```

Comment:
Reminder: should be able to use the new form of `SetStatusConditionNoHeartbeat` to replace the change check logic.